### PR TITLE
Chow liu tree structure learning

### PR DESCRIPTION
**Pull request by loudly-soft**
Implemented Chow-liu tree structure learning (Issue #1278) and unit tests.

### CODE REVIEW

**Comment by ankurankan** [comment]
@loudly-soft Quick question: Is everything from #1266 is included in this PR? If yes, I will close the older PR and go through this.

File: pgmpy/estimators/TreeAugmentedNaiveBayesSearch.py
Comment (ankurankan) [inline pgmpy/estimators/TreeAugmentedNaiveBayesSearch.py]: The imports should be in the order (with an empty line in between each category):
1. Python core imports
2. Third party package imports
3. pgmpy imports.

File: pgmpy/estimators/TreeAugmentedNaiveBayesSearch.py
Comment (ankurankan) [inline pgmpy/estimators/TreeAugmentedNaiveBayesSearch.py]: `node` should be the expected data type. pgmpy can accept any variable name which is hashable, so can instead use `hashable object`.

File: pgmpy/estimators/TreeAugmentedNaiveBayesSearch.py
Comment (ankurankan) [inline pgmpy/estimators/TreeAugmentedNaiveBayesSearch.py]: Please also add an `Examples` section with some basic usage examples.

File: pgmpy/estimators/TreeAugmentedNaiveBayesSearch.py
Comment (ankurankan) [inline pgmpy/estimators/TreeAugmentedNaiveBayesSearch.py]: The error message can be made a bit informative by also showing the given class node. So, something on the lines of: Given class node is: {class_node} but no columns in data exist by that name.

File: pgmpy/estimators/TreeAugmentedNaiveBayesSearch.py
Comment (ankurankan) [inline pgmpy/estimators/TreeAugmentedNaiveBayesSearch.py]: Same here.

File: pgmpy/estimators/TreeAugmentedNaiveBayesSearch.py
Comment (ankurankan) [inline pgmpy/estimators/TreeAugmentedNaiveBayesSearch.py]: If I am understanding this code block correctly, you are trying to compute the mutual info score between all possible combination of variables, creating an edge between them and using the mutual info as the edge weight. I think this can be simplified by computing an array of edge_weights and then creating the graph from that. Something like:
```
weights = pd.DataFrame(0, index=df_features.columns, columns=df_features.columns)
for u_index, u, v_index, v in combinations(enumerate(df_features.columns), 2):
   weights.iat[u_index, v_index] = mutual_info_score(df_features.iloc[:, u_index], df_features.iloc[:, v_index])
G = nx.from_pandas_adjacency(weights, create_using=nx.Graph)
```
I think the main benefit would be that this would be much faster because networkx's `add_node` and `add_edge` can be super slow. Also, I think this is more readable.

File: pgmpy/estimators/TreeAugmentedNaiveBayesSearch.py
Comment (ankurankan) [inline pgmpy/estimators/TreeAugmentedNaiveBayesSearch.py]: I would suggest adding a line to set a variable `root_node` at the top after argument checks and use that throughout instead of testing it later. Something like:
```
root_node = df_features.columns[0] if self.root_node is None else self.root_node
```
After that you can just replace this if else block with: `D=nx.bfs_tree(T, root_node)`

File: pgmpy/estimators/TreeAugmentedNaiveBayesSearch.py
Comment (ankurankan) [inline pgmpy/estimators/TreeAugmentedNaiveBayesSearch.py]: It would be nice if you can add the major steps of the algorithm as comments. Something like:
```
Step 1: Argument checks
Step 2: Create a fully connected graph with edge weights as the mutual info score between them.
Step 3: Get the Maximum Spanning tree and orient the edges.
```

File: pgmpy/estimators/TreeSearch.py
Comment (ankurankan) [inline pgmpy/estimators/TreeSearch.py:20]: I think TreeSearch and NaiveBayesTreeSearch can be combined together by adding some extra parameters to the `estimate` method maybe: `return_type`. It will remove boilerplate code and also make my life easier to maintain. :D

File: pgmpy/tests/test_estimators/test_TreeAugmentedNaiveBayesSearch.py
Comment (ankurankan) [inline pgmpy/tests/test_estimators/test_TreeAugmentedNaiveBayesSearch.py]: I think it would a good idea to add some more realistic tests. For example, generate data from a model and then try to learn it back the structure.

**Review by ankurankan — CHANGES_REQUESTED** [review]

**Comment by ankurankan** [comment]
@loudly-soft Sorry for the delay in reviewing this. I went through the code, everything looks good but I have added some comments mostly regarding the implementation. Please let me know if you need any help or something isn't clear.

File: pgmpy/estimators/TreeAugmentedNaiveBayesSearch.py
Comment (ankurankan) [reply pgmpy/estimators/TreeAugmentedNaiveBayesSearch.py]: You can also add a nice `tqdm` progress bar and make the computation parallel at this step.

**Comment by loudly-soft** [comment]
> @loudly-soft Quick question: Is everything from #1266 is included in this PR? If yes, I will close the older PR and go through this.

@ankurankan Oops, yes.

File: pgmpy/estimators/TreeSearch.py
Comment (loudly-soft) [reply pgmpy/estimators/TreeSearch.py:20]: I agree and I think it may call for more.  Now that I’ve implemented TAN and the Chow-Liu algorithm, I’m tempted to roll naive bayes into the mix as well.  (I know, I did say before that naive bayes is not a special case of TAN but both are special cases of FAN or ETAN.)  However, in terms of coding, they are like Lego blocks and each block can be a standalone component:

- TAN = Chow-Liu + Naive Bayes

So, my questions are:

1. Do you see Naive Bayes as belonging to “estimators” like TAN and Chow-Liu even though it doesn’t do much structure learning from data.
2. What would be the name of this generic class that can do TAN, Chow-Liu (and Naive Bayes and maybe FAN or ETAN in the future)?

File: pgmpy/estimators/TreeAugmentedNaiveBayesSearch.py
Comment (loudly-soft) [reply pgmpy/estimators/TreeAugmentedNaiveBayesSearch.py]: I’d be interested in adding documentation to `examples` but I may be a little bit short on time in the coming weeks.  In the interest of not blocking this PR, perhaps I can create a separate ticket to keep track of that task?

File: pgmpy/estimators/TreeAugmentedNaiveBayesSearch.py
Comment (loudly-soft) [reply pgmpy/estimators/TreeAugmentedNaiveBayesSearch.py]: Do you mean replace the word `node` with `hashable object`?

File: pgmpy/estimators/TreeAugmentedNaiveBayesSearch.py
Comment (loudly-soft) [reply pgmpy/estimators/TreeAugmentedNaiveBayesSearch.py]: Sorry, I'm not familiar with multi-threading in Python.  Do you mean `joblib` and the step to parallelize is:

`weights.iat[u_index, v_index] = mutual_info_score(df_features.iloc[:, u_index], df_features.iloc[:, v_index])`

I assume Pandas is thread-safe, and/or concurrent writes wouldn't cause any race condition or deadlocks to the underlying data structures backing the dataframe.

File: pgmpy/tests/test_estimators/test_TreeAugmentedNaiveBayesSearch.py
Comment (loudly-soft) [reply pgmpy/tests/test_estimators/test_TreeAugmentedNaiveBayesSearch.py]: Since TAN is generally used for classification tasks, perhaps the assert statement should measure classification errors?

The steps are:

1. Manually build a TAN graph and CPDs
2. Generate N rows of data from graph
3. Learn the TAN graph from data
4. Use the learned graph to predict from the independent variables of the data
5. Measure classification errors

I don’t suppose you have code examples along this line already that I can bootstrap from?

(I think this might go nicely into `examples` as a demo as well.)

File: pgmpy/estimators/TreeAugmentedNaiveBayesSearch.py
Comment (ankurankan) [reply pgmpy/estimators/TreeAugmentedNaiveBayesSearch.py]: We have been using something like this for node: https://github.com/pgmpy/pgmpy/blob/dev/pgmpy/base/DAG.py#L277

File: pgmpy/estimators/TreeAugmentedNaiveBayesSearch.py
Comment (ankurankan) [reply pgmpy/estimators/TreeAugmentedNaiveBayesSearch.py]: Since this is the only way for a user to understand how to use the class, I would suggest adding some basic examples before merging this. The same examples that you are adding in the tests would work too. Or else, I can also add the examples if you don't have time for it.

File: pgmpy/estimators/TreeAugmentedNaiveBayesSearch.py
Comment (ankurankan) [reply pgmpy/estimators/TreeAugmentedNaiveBayesSearch.py]: Yes, I was suggesting using `joblib`. We can run the `for` loop in parallel as there is no order requirement in this case. You can refer to: https://joblib.readthedocs.io/en/latest/parallel.html for parallelizing for loops using joblib. We will also need an argument `n_jobs=-1` in the method for the user to specify the number of threads they want to use. But this isn't a priority change, so can be done later as well.

File: pgmpy/tests/test_estimators/test_TreeAugmentedNaiveBayesSearch.py
Comment (ankurankan) [reply pgmpy/tests/test_estimators/test_TreeAugmentedNaiveBayesSearch.py]: Yes, something along those lines. But instead of testing for classification errors, I would suggest testing for smaller parts. For example, testing the learned structure or the learned parameters on a given structure. The problem with testing classification error directly (starting from a initial model/generated data) is that there are multiple steps where uncertainty can be introduced and it all can add up to give unreliable tests (or tests failing sometimes).

File: pgmpy/estimators/TreeSearch.py
Comment (ankurankan) [reply pgmpy/estimators/TreeSearch.py:20]: We already have a model class for Naive Bayes (https://github.com/pgmpy/pgmpy/blob/dev/pgmpy/models/NaiveBayes.py). So, I would suggest modifying it to be able to represent FAN models with special cases for Naive Bayes and TAN. The estimators should stay in the `estimators` dir and be completely separate from the model class, and should just return an instance of the learned model.

**Comment by loudly-soft** [comment]
@ankurankan Thanks.  I'll try to return to this after this week at the earliest.

**Comment by loudly-soft** [comment]
@ankurankan I’ve added to `examples` and they run fine with ipython but my Jupyter notebook is broken so the examples are currently missing output.

Also, looks like appveyor is choking up:

`
Build started
git clone -q https://github.com/pgmpy/pgmpy.git C:\projects\pgmpy
fatal: unable to access 'https://github.com/pgmpy/pgmpy.git/': Could not resolve host: github.com

Command exited with code 128
`

**Comment by ankurankan** [comment]
@loudly-soft Thanks for the changes. I will have a look at it. I also restarted the appveyor build, seems to be running fine now.

**Comment by loudly-soft** [comment]
Fixed my Jupyter notebook and fleshed out the code in `examples`.

FTR, the issue was [this](https://github.com/pytorch/pytorch/issues/2575).  On Ubuntu 14.04.6 LTS, downgrading scikit-learn to 0.20.3 fixed the problem.

**Comment by ankurankan** [comment]
@loudly-soft Thanks a lot again for the PR. I went through it and everything looks great. I wanted to add a commit to this PR which changes some behavior and the algorithm should be substantially faster. Could you please check that `allow edits from maintainers` is checked for this PR (https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) as I am not able to push my commit? Also, if you have some time, could you please go through my changes and see if it looks good?

Here's a summary of my changes:

1. Changes name `return_type` to `estimator_type` as the function would return a DAG in both the cases.
2. Move `return_type` and `class_node` arguments to `estimate` method. The idea is that the user would create an estimator object and then try different estimators without needing to redefine the object.
3. Adds edge_weights parameter which can accept any function to compute edge weights. 
3. Cosmetic changes: 
 - Docs within 88 chars, for better rendering on the website
 - Using f-strings everywhere, they are cleaner and much faster
 - No definition of any `self.var` in methods other than __init__. Better for readability
 - Convert `chow-liu` to a static class.
 - Removed iteration while computing weights. 
 - Move algorithm steps inside the code.

**Review by ankurankan — APPROVED** [review]

**Comment by loudly-soft** [comment]
@ankurankan Yeah sure, I'm curious about the changes you're making.  I'll have a look.

Hmm..  "Allow edits by maintainers" is already checked on this PR.

**Comment by ankurankan** [comment]
@loudly-soft Hmm, don't know why it wasn't working earlier. But got through now. Please have a look if it looks good.

File: pgmpy/estimators/TreeSearch.py
Comment (loudly-soft) [inline pgmpy/estimators/TreeSearch.py]: I'm not sure what the convention is in `pgmpy` but I wonder, for clarity, should `edge_weights` be renamed to `edge_weights_fn` or something similar to hint that a function (or function name) is expected.

File: pgmpy/estimators/TreeSearch.py
Comment (loudly-soft) [inline pgmpy/estimators/TreeSearch.py:217]: More of a question than review.  About `Parallel`, does it preserve ordering? It must be, otherwise, the indices in these 2 blocks of code would not match up.

```
pbar = combinations(data.columns, 2)
vals = Parallel(n_jobs=n_jobs, prefer="threads")(
   delayed(edge_weights)(data.loc[:, u], data.loc[:, v]) for u, v in pbar
)
```

`weights[np.triu_indices(n_vars, k=1)] = vals`

File: pgmpy/estimators/TreeSearch.py
Comment (loudly-soft) [inline pgmpy/estimators/TreeSearch.py]: Going by the docstring, should the error msg include the options 'adjusted_mutual_info' and 'normalized_mutual_info'?

File: pgmpy/estimators/TreeSearch.py
Comment (loudly-soft) [inline pgmpy/estimators/TreeSearch.py:172]: Should 'mutual_info' or one of them be default?

**Comment by loudly-soft** [comment]
@ankurankan Thanks.  I've added my 2 cents.  Pls let me know if anything is not clear.

**Review by loudly-soft — COMMENTED** [review]
@ankurankan Thanks. I've added my 2 cents. Pls let me know if anything is not clear.

File: pgmpy/estimators/TreeSearch.py
Comment (ankurankan) [reply pgmpy/estimators/TreeSearch.py]: Fixed.

File: pgmpy/estimators/TreeSearch.py
Comment (ankurankan) [reply pgmpy/estimators/TreeSearch.py:217]: Yes, it does preserve the order.

File: pgmpy/estimators/TreeSearch.py
Comment (ankurankan) [reply pgmpy/estimators/TreeSearch.py]: Yes, have changed it to `edge_weights_fn` now.

**Comment by ankurankan** [comment]
@loudly-soft Thanks for the comments and the PR.

**Comment by loudly-soft** [comment]
@ankurankan And thanks for the nifty tool again!

**Pull request merged**
